nexus/mgs-updates/src/rot_bootloader_updater.rs

```diff
@@ -265,96 +263,53 @@ impl SpComponentUpdateHelperImpl for ReconfiguratorRotBootloaderUpdater {
-                        "failed to get RoT boot info";
-                        "error" => &error
-                    );
-                    return Err(PostUpdateError::FatalError { error });
-                }
-                RotState::V3 { stage0next_error, .. } => {
-                    return Ok(stage0next_error);
-                }
-                // The RoT is probably still booting
-                RotState::CommunicationFailed { message } => {
-                    if before.elapsed() >= timeout {
-                        error!(
-                            log,
-                            "failed to get RoT boot info";
-                            "error" => %message
-                        );
-                        return Err(PostUpdateError::FatalError {
-                            error: message,
-                        });
-                    }
-
-                    info!(
-                        log,
-                        "failed getting RoT boot info (will retry)";
-                        "error" => %message,
-                    );
-                    tokio::time::sleep(WAIT_FOR_BOOT_INFO_INTERVAL).await;
-                }
-            },
-            // The RoT might still be booting
-            Err(error) => {
-                let e = InlineErrorChain::new(&error);
-                if before.elapsed() >= timeout {
-                    error!(
-                        log,
-                        "failed to get RoT boot info";
-                        &e,
-                    );
-                    return Err(PostUpdateError::FatalError {
-                        error: e.to_string(),
-                    });
-                }
-
-                info!(
+    match wait_for_boot_info(log, mgs_clients, sp_type, sp_slot, timeout).await
+    {
+        Ok(state) => match state {
+            // The minimum we will ever return is v3.
+            // Additionally, V2 does not report image errors, so we cannot
+            // know with certainty if a signature check came back with errors
+            RotState::V2 { .. } => {
+                let error = "unexpected RoT version: 2".to_string();
+                error!(
                     log,
-                    "failed getting RoT boot info (will retry)";
-                    e,
+                    "failed to get RoT boot info";
+                    "error" => &error
                 );
-                tokio::time::sleep(WAIT_FOR_BOOT_INFO_INTERVAL).await;
+                return Err(PostUpdateError::FatalError { error });
             }
-        }
+            RotState::V3 { stage0next_error, .. } => {
+                return Ok(stage0next_error);
+            }
+            // This is unreachable because wait_for_boot_info loops for some
+            // time if it encounters `CommunicationFailed`, and if it hits the
+            // timeout, it will return an error.
+            RotState::CommunicationFailed { .. } => unreachable!(),
```

Comment:
This is correct but makes me a little nervous because it'd depending on behavior from a function in another file; if I'm changing that function there's nothing that forces me to realize this dependency exists. It looks like the only `RotState` variant we can currently deal with is `RotState::V3`, right? I wonder if it would make sense for `wait_for_boot_info` to return its own enum (names are hard, but maybe something like `CheckedRotState` or `RotBootInfo` or something like that) that only has a `V3` variant? Then it could return errors for V2 just like this function does, and it could omit `CommunicationFailed` entirely since it handles it internally (which means we wouldn't need this `unreachable!()`).

Reply:
Yeah, fair. I guess I wish we had `RotState::V3(RotStateV3)` where `RotStateV3` was an explicitly-defined struct; then you could reuse that instead of having to duplicate all the fields. But that's a bit of work that spans multiple repos; just returning an error here seems fine.